### ISSUE TITLE
a11y violations appearing on the IBM Dynamic Assessment Plugin (DAP)

### ISSUE THREAD

**Issue by arbydev** [carbon-design-system/carbon#341]
I ran the Dynamic Assessment Plugin (DAP)
https://www-03.ibm.com/able/dynamic-assessment-plug-in.html
on multiple pages throughout http://carbondesignsystem.com 
These are some of the violations it returned:

Each form control must have associated label.
Multiple WAI-ARIA 'search' landmarks must have unique labels specified with aria-label or aria-labelledby.
The WAI-ARIA widget with img role must have a nonempty alternate text or a label.
The WAI-ARIA role(s) and/or attribute(s) identifier are not valid for the element identifier. See rule help for guidance.
Minimum contrast of identifier with its background is not sufficient to meet WCAG AA requirements for text of size identifierpx and weight of identifier.
Multiple WAI-ARIA 'navigation' landmarks must have unique labels specified with aria-label or aria-labelledby.

Could these be addressed for Accessibility Verification Testing (AVT)?

**Comment by carmacleod** [comment]
@aagonzales @marijohannessen If you want to grab some time in my calendar, I can walk you through running the DAP on your example/doc pages. I recognize that many/most of the pages on http://carbondesignsystem.com are "just doc", however in order to truly claim that the Carbon Components are [accessible](http://www.carbondesignsystem.com/guidelines/accessibility), the doc needs to be accessible as well.  :)  (We could even beef up that Accessibility Guidelines page a bit as well - maybe mention landmarks, arrow/esc keys, dialogs, etc, etc).

In addition, every component on all example pages should be accessible, in order to show best practices when using Carbon Components. Ideally, we could point out the things that consumers of each component need to do in order to achieve accessibility (make sure there's a textual label, etc, etc).

If you have a bit of money in the budget, you could also contract an AVT team to run Accessibility Verification Testing for you - they will test using the DAP, the keyboard, and the JAWS screen reader on Windows/Firefox. They can also run VoiceOver on Safari/Mac OS X.  I can recommend a team for you - just ask.

**Comment by aagonzales** [comment]
@carmacleod @arbydev Could you please point specific cases/instances where things are failing. We have also made updates since this issue was created, are this issues still existing?

We always check for accessibility and it is a priority for us so blanket statements like this can be confusing for us without specifics. We also work closely and have regular meeting with Bo Campbell from IBM Accessibility.

> We could even beef up that Accessibility Guidelines page a bit as well - maybe mention landmarks, arrow/esc keys, dialogs, etc, etc).
- Yes, we would love to do this. We've had this in our backlog for awhile. We'd love any help on this.

**Comment by carmacleod** [comment]
Oh! Apologies, @aagonzales - I just happened to notice this open issue - I didn't actually run the tools to see for myself. Mea culpa.

So, running the DAP very quickly on one page (http://carbondesignsystem.com), I see the following:
- the Search needs a label (either a visible label element, or at a minimum, an invisible aria-label="Search:"  (because [placeholder does not accessibly replace a label](https://www.w3.org/WAI/tutorials/forms/instructions/#placeholder-text))
- the Getting Started, Style, Components, and Resources images have alt text that is redundant with the adjacent link text. These images are purely decorative, so they need to have alt="" (instead of, for example, alt="Getting started"). This is so that screen readers don't read the text out twice.
- there are 2 `<main>` elements. I would suggest that the one with id="root" should just be a div, in which case it wouldn't need the `role="main" aria-label="Main website wrapper"`. The other one, with id="maincontent", currently has aria-labelledby="page-title" but there's no id="page-title" on the page. That should either point to a real id, or be removed; main elements don't really need a label because [there's only supposed to be one (visible) main element](http://w3c.github.io/html/grouping-content.html#the-main-element).
- If you remove the main wrapper to fix the "2 mains not allowed" problem, above, then you will be notified that the "Toggle Side Nav" button has been "orphaned", i.e. it is not inside a landmark region. To fix this, please move the toggle side nav button into the `<nav>` somehow - maybe the
`<div data-reactroot class="wrapper">` could be the `<nav>` instead, or maybe a new element could wrap the button and the current nav, and the current nav could be demoted to div. ;)
- the search input does not need role="search" because its parent div (the one with class="bx--search bx--search--sm") already has role="search".
- there are some SVG's with `role="img" aria-label="Menu arrow icon"` and `aria-hidden="true"`. Please note that aria-hidden="true" is ignored when an element has a role and a label. So for these, I think you want to remove the aria-label and the aria-hidden, and just use `role="presentation"` (which removes them from the browser's accessibility tree) because I think these images are purely presentational. (Alternatively, you can use `role="none"`, which is synonymous with `role="presentation"` but "none" only works on newer browsers).
- similarly with the SVG's with `aria-label="sidenav link icon"` - just use `role="presentation"` (not "img") and remove the aria-label, because these arrows are purely presentational and they don't need to be exposed to AT.
- the 4 [section elements](http://w3c.github.io/html/sections.html#the-section-element) in the main need to either have a label (either with a visible heading pointed to by aria-labelledby, or with an invisible aria-label), or they need to be turned into divs. This is because `<section>` elements are supposed to be able to be used in the document outline, and so they need a label. Also, screen readers can navigate to sections, but the user won't know which section they are navigating to unless it has a label. I think these 4 sections could be marked up as follows:
```
<header class="overview-page__banner"> ...  (i.e. change this one to a header (which is a banner landmark) instead of a section; then it doesn't need an aria-label because there's only one banner per page. It would be a tiny bit unexpected to have the banner landmark _inside_ the main landmark, so if possible, could the banner be moved outside the main? i.e. just in the markup - it would still look the same)
<section class="overview-page__section" aria-label="Overview"> ...
<section class="overview-page__section overview-page__section--gray" aria-label="Code"> ... (or maybe "Working with Carbon"? ... not quite sure what the label for this section would be, but not "gray" :)
<section class="overview-page__section overview-page__section--contribute" aria-label="Contribute" ...> ...
```

In case it's helpful, here's a great explanation of [HTML5 section elements and ARIA landmark regions](https://www.w3.org/TR/wai-aria-practices/examples/landmarks). (Click on items in the left nav to read about them; click on show/hide landmarks/headings to see those).

**Comment by marijohannessen** [comment]
Issue moved to [carbon-design-system/design-system-website #152](https://github.com/carbon-design-system/design-system-website/issues/152) via [**ZenHub**](https://www.zenhub.com/)
